gollm/openai.go

```diff
@@ -48,26 +73,61 @@ type OpenAIClient struct {
 // Ensure OpenAIClient implements the Client interface.
 var _ Client = &OpenAIClient{}
 
+// getOpenAIModel returns the appropriate model based on configuration and explicitly provided model name
+// Priority order for model selection:
+// 1. Explicitly provided model parameter (highest)
+// 2. CLI flag --model
+// 3. KUBECTL_AI_MODEL environment variable
+// 4. OPENAI_MODEL environment variable (via binding in init())
+// 5. Default value "gpt-4.1" (lowest)
+func getOpenAIModel(model string) string {
+	// If explicit model is provided, use it
+	if model != "" {
+		klog.V(2).Infof("Using explicitly provided model: %s", model)
+		return model
+	}
+
+	// Check configuration
+	configModel := viper.GetString("model")
+	if configModel != "" {
+		klog.V(1).Infof("Using model from config: %s", configModel)
+		return configModel
+	}
+
+	// Default model as fallback
+	klog.V(2).Info("No model specified, defaulting to gpt-4.1")
+	return "gpt-4.1"
+}
+
 // NewOpenAIClient creates a new client for interacting with OpenAI.
-// It reads the API key and optional endpoint from environment variables
-// OPENAI_API_KEY and OPENAI_ENDPOINT.
+// It reads the API key and optional endpoint from viper configuration.
+// For backward compatibility, it also supports OPENAI_API_KEY and
+// OPENAI_ENDPOINT environment variables through viper bindings.
 func NewOpenAIClient(ctx context.Context) (*OpenAIClient, error) {
-	apiKey := os.Getenv("OPENAI_API_KEY")
+	// Get API key from viper configuration
+	// This will also check the OPENAI_API_KEY env var due to the binding in init()
+	apiKey := viper.GetString("openai-api-key")
 	if apiKey == "" {
-		// The NewClient might handle this, but explicit check is safer
-		return nil, errors.New("OPENAI_API_KEY environment variable not set")
+		return nil, errors.New("OpenAI API key not found. Set via OPENAI_API_KEY, KUBECTL_AI_OPENAI_API_KEY env var, or openai-api-key in config")
 	}
 
-	endpoint := os.Getenv("OPENAI_ENDPOINT")
-	if endpoint != "" {
-		klog.Infof("Using custom OpenAI endpoint: %s", endpoint)
-		return &OpenAIClient{
-			client: openai.NewClient(option.WithBaseURL(endpoint)),
-		}, nil
+	// Set options for client creation
+	options := []option.RequestOption{option.WithAPIKey(apiKey)}
+
+	// Check for custom endpoint or API base URL
+	// Endpoint takes precedence over API base if both are defined
+	baseURL := viper.GetString("openai-endpoint")
+	if baseURL == "" {
+		baseURL = viper.GetString("openai-api-base")
+	}
+
+	if baseURL != "" {
+		klog.Infof("Using custom OpenAI base URL: %s", baseURL)
+		options = append(options, option.WithBaseURL(baseURL))
```

Comment:
My bad, I should have stated this after we merged the cobra integration, but I am reconsidering the decision to integrate viper in `kubectl-ai`. Sorry to kill all the excitement here :(

Input to the CLI programs are like APIs. Once introduced, we have to maintain them forever. I am finding that viper is expanding that surface area too quickly (for ex. all these environment variable for all the flags...) which may become problem for us later.

I am open to introducing multiple ways to configure the program but slowly overtime based on the need.

I am finding viper is introducing too many dependency and complexity so soon in the project with very little value.

@tuannvm I hope you understand. If needed, happy to connect over chat (or call).

Reply:
https://github.com/GoogleCloudPlatform/kubectl-ai/pull/214